### PR TITLE
HTTP/2 implementation: HTTP/2-cache (#309).

### PR DESCRIPTION
**Pull request by aleksostapenko**

### CODE REVIEW

File: tempesta_fw/http.h [hunk 1]
```diff
@@ -614,6 +618,33 @@ tfw_http_resp_code_range(const int n)
...
 }
 
+/*
+ * Static index determination for response ':status' pseudo-header (see RFC
+ * 7541 Appendix A for details).
+ */
+static inline unsigned short
+tfw_h2_pseudo_index(unsigned short status)
+{
+	switch (status) {
```
Comment (nmmmnu) [inline tempesta_fw/http.h]: Wouldn't this be faster wit array?
Of course will use minimum 512 bytes.

Comment (krizhanovsky) [reply tempesta_fw/http.h]: The compiler will do binary search over the labels which is much faster than accessing 4 cache line array. The compiler also generates lookup table (exactly the array you proposed), but for more localized values.

File: tempesta_fw/http_msg.c [hunk 2]
```diff
@@ -168,6 +168,13 @@ tfw_http_msg_req_spec_hid(const TfwStr *hdr)
 /**
  * Fills @val with second part of special HTTP header containing the header
  * value.
+ *
+ * TODO: this function should be brought to a uniform look with other similar
+ * procedures( @tfw_http_hdr_split() and @tfw_h2_msg_hdr_length()), more
+ * likely - aggregated into one general-purpose procedure, after HTTP/1.1-parser
+ * extending (see also TODO-comment for @tfw_http_hdr_split_cp() procedure); for
+ * now this procedure leaves RWS in the resulting @val and this is not correct
+ * in context of HTTP/2<=>HTTP/1.1 comparisons, transformations etc.
```
Comment (krizhanovsky) [inline tempesta_fw/http_msg.c]: Just leave a comment to remember that the TODO must be fixed in upcoming PR.

File: tempesta_fw/http.c [hunk 3]
```diff
@@ -3688,26 +3895,78 @@ tfw_http_hdr_split(TfwStr *hdr, TfwStr *name_out, TfwStr *val_out)
...
+}
+
+/*
+ * Split header in two parts: name and value, with descriptors copying (see
+ * description of @tfw_http_hdr_split() for details).
+ *
+ * TODO: this procedure should be evicted after HTTP/1.1-parser extending
+ * (and @tfw_http_hdr_split()/@tfw_h2_msg_hdr_length procedures upgrading) in
+ * order to split headers into separate chunks for name, ':', LWS, value, and
+ * RWS - during the HTTP-message parsing stage.
```
Comment (krizhanovsky) [inline tempesta_fw/http.c]: Just leave a comment to remember that the TODO must be fixed in upcoming PR.

File: tempesta_fw/cache.c [hunk 4]
```diff
@@ -936,73 +1086,181 @@ tfw_cache_strcpy_eol(char **p, TdbVRec **trec,
...
+				return -ENOMEM;
+
+			st_index = dup->hpack_idx;
+			CSTR_MOVE_HDR();
+		}
+
+		if (st_index) {
+			if (tfw_cache_h2_copy_int(&ce->hdr_len, st_index, 0xF,
+						  p, trec, tot_len))
+				return -ENOMEM;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: Yes - it is exactly how it works now: the response from cache is served through `tfw_http_req_cache_service()` -> `tfw_h2_resp_fwd()` flow, bypassing the `tfw_h2_resp_adjust_fwd()` procedure.

File: tempesta_fw/cache.c [hunk 5]
```diff
@@ -1173,58 +1466,109 @@ tfw_cache_copy_resp(TfwCacheEntry *ce, TfwHttpResp *resp, size_t tot_len)
...
-			h = &empty;
-		} else {
+		int hid = field - resp->h_tbl->tbl;
+		/*
+		 * Skip hop-by-hop headers. Also skip 'Server' header (with
+		 * possible duplicates), since we will substitute it with our
+		 * version of this header.
+		 */
+		if ((field->flags & TFW_STR_HBH_HDR)
+		    || hid == TFW_HTTP_HDR_SERVER
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: We must preserve all the already existing `Via` headers in the message while forwarding it (https://tools.ietf.org/html/rfc7230#section-5.7.1). The resulting order of the `Via` headers remains correct, as we add our `Via` header in the end - after all the response headers are already copied.

File: tempesta_fw/http.c [hunk 6]
```diff
@@ -3871,71 +3905,33 @@ tfw_http_hdr_split(TfwStr *hdr, TfwStr *name_out, TfwStr *val_out)
...
-		tail = 0;
-		idx = chunk->len - 1;
-		while (chunk->data[idx] == ' '
-		       || chunk->data[idx] == '\t')
-		{
-			++tail;
-			if (unlikely(!idx))
-				break;
-			--idx;
-		}
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Yeah - there is no need to iterate characters inside the chunks any more.

File: tempesta_fw/http_sess.c [hunk 7]
```diff
@@ -460,8 +454,21 @@ tfw_http_sticky_add(TfwHttpResp *resp)
...
 		set_cookie.hpack_idx = 55;
-		r = __hdr_h2_add(resp, &set_cookie);
+		r = tfw_hpack_encode(resp, &set_cookie, op, !cache);
+	}
+	else if (cache) {
+		TfwHttpTransIter *mit = &resp->mit;
+		struct sk_buff **skb_head = &resp->msg.skb_head;
+		TfwStr crlf = { .data = S_CRLF, .len = SLEN(S_CRLF) };
+
+		r = tfw_http_msg_expand_data(&mit->iter, skb_head,
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http_sess.c]: The problem here is that along with copying headers from cache and addition the known in advance headers (e.g. `set-cookie`, `date` etc.) we must also process configured headers corrections (which can be addition/substitution/deletion/appending) and we cannot predict here the resulting size of memory for the skbs to be allocated in the target response.

File: tempesta_fw/cache.c [hunk 8]
```diff
@@ -1651,23 +2146,35 @@ cache_req_process_node(TfwHttpReq *req, tfw_http_cache_cb_t action)
...
 	 */
-	if (tfw_cache_set_hdr_age((TfwHttpMsg *)resp, ce)) {
-		T_WARN("Unable to add Age: header, cached response [%p] "
-		       "dropped\n", resp);
-		TFW_INC_STAT_BH(clnt.msgs_otherr);
-		tfw_http_msg_free((TfwHttpMsg *)resp);
-		resp = NULL;
-		goto out;
+	if (TFW_MSG_H2(req)) {
+		id = tfw_h2_stream_id(req);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: >stream can't be destroyed until request is alive

Actually can: during processing of request on http layer, the frame layer can receive the RST_STREAM from the client and stream will be moved to the final `closed` state (and consequently - will be removed).

[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/cache.c]: Ok, i will check that, but it seems to be resource management question then. Stream can be removed, but stream object must exist until referencing request exists. Same as http1.1 requests keep `connection` object alive. 

Anyway. no need to change this right now, i was curious for locking reason.

File: tempesta_fw/cache.c [hunk 9]
```diff
@@ -936,73 +1086,170 @@ tfw_cache_strcpy_eol(char **p, TdbVRec **trec,
...
+		unsigned long h_len;
+
+		TFW_STR_INIT(&s_nm);
+		TFW_STR_INIT(&s_val);
+		tfw_http_hdr_split(hdr, &s_nm, &s_val, true);
+
+		st_index = hdr->hpack_idx;
+		h_len = tfw_h2_hdr_size(s_nm.len, s_val.len, st_index);
+
+		/* Don't split short stprings. */
```
Comment (ikoveshnikov) [inline tempesta_fw/cache.c]:
```suggestion
		/* Don't split short strings. */
```

File: tempesta_fw/cache.c [hunk 10]
```diff
@@ -562,56 +576,154 @@ tfw_cache_write_field(TDB *db, TdbVRec **trec, TfwHttpResp *resp,
...
 	BUG_ON(*p > (*trec)->data + (*trec)->len);
 
-	if (likely(!(s->flags & TFW_STR_DUPLICATE)))
-		return tfw_cache_write_field(db, trec, resp, it, p, s->len,
-					     hdr);
+	if (likely(!(s->flags & TFW_STR_DUPLICATE))) {
+		r = write_actor(db, trec, resp, p, s->len, &dc_iter);
+		if (likely(!r))
+			*acc_len += dc_iter.acc_len;
+		goto out;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: Corrected.

File: tempesta_fw/http.c [hunk 11]
```diff
@@ -3688,26 +3895,78 @@ tfw_http_hdr_split(TfwStr *hdr, TfwStr *name_out, TfwStr *val_out)
...
+int
+tfw_h2_resp_add_loc_hdrs(TfwHttpResp *resp, const TfwHdrMods *h_mods,
+			 bool cache)
 {
 	unsigned int i;
 	TfwHttpTransIter *mit = &resp->mit;
+	TfwH2TransOp op = cache ? TFW_H2_TRANS_EXPAND : TFW_H2_TRANS_ADD;
+
+	if (!h_mods)
+		return 0;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/cache.c [hunk 12]
```diff
@@ -1529,44 +1978,33 @@ tfw_cache_build_resp_body(TDB *db, TfwHttpResp *resp, TdbVRec *trec,
  * TODO use iterator and passed skbs to be called from net_tx_action.
  */
 static TfwHttpResp *
-tfw_cache_build_resp(TfwHttpReq *req, TfwCacheEntry *ce)
+tfw_cache_h2_build_resp(TfwHttpReq *req, TfwCacheEntry *ce, time_t lifetime,
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: Corrected (reverted names).

File: tempesta_fw/cache.c [hunk 13]
```diff
@@ -562,56 +576,154 @@ tfw_cache_write_field(TDB *db, TdbVRec **trec, TfwHttpResp *resp,
...
+		BUG_ON(!tr);
+		*data = tr->data;
+	}
+
+	return r;
+}
+
+static inline int
+tfw_cache_h2_set_status(TDB *db, TfwCacheEntry *ce, TfwHttpResp *resp,
+			TdbVRec **trec,	char **p, unsigned long *acc_len)
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: Corrected.

File: tempesta_fw/cache.c [hunk 14]
```diff
@@ -1173,58 +1455,108 @@ tfw_cache_copy_resp(TfwCacheEntry *ce, TfwHttpResp *resp, size_t tot_len)
...
-			h = &empty;
-		} else {
+		int hid = field - resp->h_tbl->tbl;
+		/*
+		 * Skip hop-by-hop headers. Also skip 'Server' header (with
+		 * possible duplicates), since we will substitute it with our
+		 * version of this header.
+		 */
+		if ((field->flags & TFW_STR_HBH_HDR)
+		    || hid == TFW_HTTP_HDR_SERVER
```
Comment (ikoveshnikov) [inline tempesta_fw/cache.c]: Let's skip `Set-Cookie` header. Actually backend should protect responses with private cookies from caching, but I'll be good not to spread the same cookie among all clients if it mistakenly leaks. What do you think?

File: tempesta_fw/http_msg.c (hunk 2 above)
Comment (aleksostapenko) [reply tempesta_fw/http_msg.c]: TODO is fixed.

File: tempesta_fw/http.c (hunk 3 above)
Comment (aleksostapenko) [reply tempesta_fw/http.c]: TODO is fixed.

File: tempesta_fw/http.h [hunk 15]
```diff
@@ -302,6 +300,9 @@ enum {
 	((!hmmsg->conn || TFW_CONN_TYPE(hmmsg->conn) & Conn_Srv) &&	\
 	 hmmsg->pair && TFW_MSG_H2(hmmsg->pair))
 
+#define H2_STAT_VAL_LEN			3
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.h]: Done: moved to `http_msg.h` with comment.

File: tempesta_fw/http.h [hunk 16]
```diff
@@ -302,6 +300,9 @@ enum {
...
+#define RESP_BUF_LEN			128
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.h]: Done.

File: tempesta_fw/cache.c (hunk 9 above)
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: Corrected.

File: tempesta_fw/cache.c (hunk 14 above)
Comment (aleksostapenko) [reply tempesta_fw/cache.c]: As a result of discussion in the chat, it was decided to postpone this change and additionally investigate it.

Comment (ikoveshnikov) [reply tempesta_fw/cache.c]: Made some research and moved the requirement to #1075 which is all about cookies and security.

**Pull request merged**